src/plone/restapi/services/content/utils.py

```diff
@@ -22,17 +28,56 @@ def create(container, type_, id_=None, title=None):
             str(now.millis())[7:],
             randint(0, 9999))
     else:
-        new_id = id_
+        if isinstance(id_, unicode):
+            new_id = id_.encode('utf8')
+        else:
+            new_id = id_
+
+    portal_types = getToolByName(container, 'portal_types')
+    type_info = portal_types.getTypeInfo(type_)
+
+    # Check for add permission
+    if not type_info.isConstructionAllowed(container):
+        raise Unauthorized('Cannot create %s' % type_info.getId())
+
+    # Check if allowed subobject type
+    container_type_info = portal_types.getTypeInfo(container)
+    if not container_type_info.allowType(type_):
+        raise Unauthorized('Disallowed subobject type: %s' % type_)
+
+    # Check for type constraints
+    if type_ not in [fti.getId() for fti in container.allowedContentTypes()]:
+        raise Unauthorized('Disallowed subobject type: %s' % type_)
 
-    try:
-        new_id = container.invokeFactory(type_, new_id, title=title)
-    except (BadRequest, ValueError) as exc:
-        return {'error': {
-            'type': exc.__class__.__name__,
-            'message': exc.message,
-        }}
+    if type_info.product:
+        # Oldstyle factory
+        factory = type_info._getFactoryMethod(container, check_security=0)
+        new_id = factory(new_id, title=title)
+        obj = container._getOb(new_id)
 
-    return container[new_id]
+    else:
+        factory = getUtility(IFactory, type_info.factory)
+        obj = factory(new_id, title=title)
+
+    if base_hasattr(obj, '_setPortalTypeName'):
+        obj._setPortalTypeName(type_info.getId())
+
+    return obj
+
+
+def add(container, obj):
+    """Add an object to a container."""
+    id_ = obj.getId()
+    # Archetypes objects are already created in a container thus we just fire
+    # the notification events.
+    if container == aq_parent(obj):
```

Comment:
Comparing acquisition-wrapped objects with `==` can be problematic (two different wrappers of the same object are not equal). Probably best to do `if aq_base(container) is aq_base(aq_parent(obj))`

And maybe we should raise an exception if obj already has a different parent?

Reply:
Thanks, I've amended that.

Not sure, how a different parent would be possible at all? Have you something in mind how this could happen?